### PR TITLE
New feature: LiDAR-based panoptic segmentation - EfficientLPS

### PR DESCRIPTION
**Pull request by aselimc**
Hi,

This PR adds [EfficientLPS](http://lidar-panoptic.cs.uni-freiburg.de/) network. The original repo can be found [here](https://github.com/robot-learning-freiburg/EfficientLPS). 

ROS Nodes are also added together with this PR. Furthermore, we have introduced PointCloud2 publisher, which is the latest PointCloud message type in ROS.

This PR does not need to be merged before the next release.

[Edit] Performance evaluation has to be completed but I don't have a recent Jetson right now. @vniclas

### CODE REVIEW

**Review by aselimc — COMMENTED** [review]
Sorry, here there is a typo, it should be "install.sh"

File: tests/sources/tools/perception/panoptic_segmentation/efficient_lps/test_efficient_lps.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
+# Copyright 2020-2022 OpenDR European Project
```
Comment (vniclas) [inline tests/sources/tools/perception/panoptic_segmentation/efficient_lps/test_efficient_lps.py:1]: Please update to 2023

File: projects/python/perception/panoptic_segmentation/efficient_lps/example_usage.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+import os
+import sys
+from pathlib import Path
+
+from opendr.perception.panoptic_segmentation import EfficientLpsLearner, SemanticKittiDataset
+
+# User must download the datasets and set the paths accordingly
+DATA_ROOT = '/home/USER/data/efficientLPS'
+SEMANTIC_KITTI_ROOT = f'{DATA_ROOT}/converted_datasets/semantickitti_panoptic'
```
Comment (ad-daniel) [inline projects/python/perception/panoptic_segmentation/efficient_lps/example_usage.py:23]: does it have to be these locations? If not then perhaps then the data root should be passed instead by argument and default somewhere relative to the demo itself

File: src/opendr/perception/panoptic_segmentation/efficient_lps/efficient_lps_learner.py [hunk 3]
```diff
@@ -0,0 +1,829 @@
...
+        proj = (proj - sensor_img_means
+                ) / sensor_img_stds
+        proj = proj * proj_mask.float()
+
+        mmdet_pcl = {
+            # TODO: Improve Awful Hack!
+            # Prediction will be saved in $PWD/tmpDir/inference/predictions/pcl_<file_count>.label
+            #                                         |||||||||             \\\\\\\\\\\\\\\\\
+            #                                         VVVVVVVVV              VVVVVVVVVVVVVVVVV
+            "filename": f"expected_path_for_something/inference/__dont_care__/pcl_{frame_id:06d}.bin",
```
Comment (ad-daniel) [inline src/opendr/perception/panoptic_segmentation/efficient_lps/efficient_lps_learner.py:415]: can it be improved? :smile:

File: tests/sources/tools/perception/panoptic_segmentation/efficient_lps/test_efficient_lps.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    def test_semantic_kitti_dataset(self):
+        # Test the iterator methods of the dataset
+        # Test commented out, since it depends on having the full Kitti Dataset installed somewhere
+        # kitti_path = "~/dat/kitti/dataset/"
+        #
+        # for split in ["train", "valid", "test"]:
+        #     dataset = SemanticKittiDataset(kitti_path, split=split)
+        #
+        #     # Test the length of the dataset
+        #     len_dataset = len(dataset)
```
Comment (ad-daniel) [inline tests/sources/tools/perception/panoptic_segmentation/efficient_lps/test_efficient_lps.py:174]: should be deleted if not needed

**Review by ad-daniel — CHANGES_REQUESTED** [review]
Thank you, just some comments. Also you should
- add a changelog entry
- add a link to the demo in index.md

File: projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+            time.sleep(0.1)
+
+            i += 1
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    # I have defined the default path in the place where I have my example data, this should be adjusted later on
+    parser.add_argument('dataset_path', type=str, default='/home/canakcia/datasets/semantickitti/test_data',
```
[the remark this answers is not in the document]
Comment (aselimc) [reply projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py:68]: Hey, thanks for the review. I have a short question regarding this default path. Is it sufficient enough if the default path is written something like "/home/USER/datasets/semantickitti/" or do you have other suggestions?

File: projects/opendr_ws/src/opendr_bridge/src/opendr_bridge/bridge.py [hunk 6]
```diff
@@ -599,6 +602,75 @@ def to_ros_point_cloud(self, point_cloud: PointCloud):
 
         return ros_point_cloud
 
+    def from_ros_point_cloud2(self, point_cloud: PointCloud2Msg):
```
[the remark this answers is not in the document]
Comment (aselimc) [reply projects/opendr_ws/src/opendr_bridge/src/opendr_bridge/bridge.py:605]: Please have a look at the commit  9f96820 for review. Thanks.

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 7]
```diff
@@ -354,6 +354,119 @@ whose documentation can be found here: [Deep Sort docs](../../../../docs/referen
...
+rosrun perception panoptic_segmentation_efficient_lps.py /opendr/dataset_point_cloud2
+```
+
+The following optional arguments are available:
+- `-h, --help`: show a help message and exit
+- `--input_pcl_topic INPUT_PCL_TOPIC` : listen to pointclouds on this topic (default=`/usb_cam/pcl_raw`)
+- `--checkpoint CHECKPOINT` : download pretrained models [semantickitti, nuscenes (future)] or load from the provided path (default=`semantickitti`)
+- `--output_rgb_visualization_topic OUTPUT_RGB_VISUALIZATION_TOPIC`: publish the rgb visualization on this topic on `OUTPUT_RGB_VISUALIZATION_TOPIC` (default=`/opendr/panoptic`) 
+
+### Semantic Segmentation ROS Node
```
[the remark this answers is not in the document]
Comment (aselimc) [reply projects/opendr_ws/src/opendr_perception/README.md:391]: Yes this is a mistake coming from merge conflict resolution. I tried to revert this file back to original develop version and just added the necessary changes. Please have a look at [59d0778](https://github.com/opendr-eu/opendr/pull/359/commits/59d07783384e17432bcca928e62f0f2e359e7d93). Thanks!

File: projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ad-daniel) [reply projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py:68]: I think it should be sufficient something like: `./datasets/semantickitti`
Then you should check if the folder exists, and if not it should be downloaded there (assuming that's an option), or print a warning about it so that the user downloads it himself

[the remark this answers is not in the document]
Comment (ad-daniel) [reply projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py:68]: also now that I noticed it, for the parser argument should be `--dataset_path` rather than `dataset_path` and perhaps add a short version too with `-d`

File: projects/python/perception/panoptic_segmentation/efficient_lps/example_usage.py (hunk 2 above)
Comment (aselimc) [reply projects/python/perception/panoptic_segmentation/efficient_lps/example_usage.py]: I have changed this into optional argument as suggested.

File: src/opendr/perception/panoptic_segmentation/efficient_lps/efficient_lps_learner.py (hunk 3 above)
Comment (aselimc) [reply src/opendr/perception/panoptic_segmentation/efficient_lps/efficient_lps_learner.py]: I have tried to improve this by changing the original repository. Now the filename is not used at all since the 3D predictions are not saved/loaded locally but instead values are passed as variables directly. I think this should do the trick

**Comment by aselimc** [comment]
@ad-daniel , all the changes you have requested should be addressed now. Can you please re-review it ?

**Review by ad-daniel — APPROVED** [review]
Thank you! I'll take your word on the demo (80 GB :smile:)

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 8]
```diff
@@ -354,35 +354,39 @@ whose documentation can be found here: [Deep Sort docs](../../../../docs/referen
 An [image dataset node](#image-dataset-ros-node) is also provided to be used along these nodes.
 Make sure to change the default input topic of the tracking node if you are not using the USB cam node.
 
-### Panoptic Segmentation ROS Node
-
-You can find the panoptic segmentation ROS node python script [here](./scripts/panoptic_segmentation_efficient_ps_node.py) to inspect the code and modify it as you wish to fit your needs.
-The node makes use of the toolkit's [panoptic segmentation tool](../../../../src/opendr/perception/panoptic_segmentation/efficient_ps/efficient_ps_learner.py) whose documentation can be found [here](../../../../docs/reference/efficient-ps.md)
-and additional information about Efficient PS [here](../../../../src/opendr/perception/panoptic_segmentation/README.md).
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/README.md:361]: Please re-add this section and make any modifications on top of it if you wish. This is a standard format for all nodes in this document that contains all relevant links. Refer to other nodes' sections.

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 9]
```diff
@@ -354,35 +354,39 @@ whose documentation can be found here: [Deep Sort docs](../../../../docs/referen
...
-    ```
-
-    The following optional arguments are available:
-   - `-h or --help`: show a help message and exit
-   - `-i or --input_rgb_image_topic INPUT_RGB_IMAGE_TOPIC` : listen to RGB images on this topic (default=`/usb_cam/image_raw`)
-   - `-oh --output_heatmap_topic OUTPUT_HEATMAP_TOPIC`: publish the semantic and instance maps on this topic as `OUTPUT_HEATMAP_TOPIC/semantic` and `OUTPUT_HEATMAP_TOPIC/instance`, `None` to stop the node from publishing on this topic (default=`/opendr/panoptic`)
-   - `-ov --output_rgb_image_topic OUTPUT_RGB_IMAGE_TOPIC`: publish the panoptic segmentation map as an RGB image on this topic or a more detailed overview if using the `--detailed_visualization` flag, `None` to stop the node from publishing on this topic (default=`opendr/panoptic/rgb_visualization`)
-   - `--detailed_visualization`: generate a combined overview of the input RGB image and the semantic, instance, and panoptic segmentation maps and publish it on `OUTPUT_RGB_IMAGE_TOPIC` (default=deactivated)
-   - `--checkpoint CHECKPOINT` : download pretrained models [cityscapes, kitti] or load from the provided path (default=`cityscapes`)
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/README.md:379]: Similar to the previous comment, re-add these deleted lines and make any modification on top if you think it's necessary. The "instructions for basic usage" section is standard for all nodes in the document.

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 10]
```diff
@@ -354,35 +354,39 @@ whose documentation can be found here: [Deep Sort docs](../../../../docs/referen
...
+### Vision Based Panoptic Segmentation ROS Node
+A ROS node for performing panoptic segmentation on a specified RGB image stream using the [EfficientPS](../../../../src/opendr/perception/panoptic_segmentation/README.md) network.
+Assuming that the OpenDR catkin workspace has been sourced, the node can be started with:
+```shell
+rosrun opendr_perception panoptic_segmentation_efficient_ps.py
+```
 
-3. Default output topics:
-   - Output images: `/opendr/panoptic/semantic`, `/opendr/panoptic/instance`, `/opendr/panoptic/rgb_visualization`
-   - Detection messages: `/opendr/panoptic/semantic`, `/opendr/panoptic/instance`
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/README.md:383]: Similar to previous comment, re-add this.

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 11]
```diff
@@ -354,35 +354,39 @@ whose documentation can be found here: [Deep Sort docs](../../../../docs/referen
...
+```shell
+rosrun opendr_perception panoptic_segmentation_efficient_lps.py /opendr/dataset_point_cloud2
+```
 
-   For viewing the output, refer to the [notes above.](#notes)
+The following optional arguments are available:
+- `-h, --help`: show a help message and exit
+- `--input_pcl_topic INPUT_PCL_TOPIC` : listen to pointclouds on this topic (default=`/usb_cam/pcl_raw`)
+- `--checkpoint CHECKPOINT` : download pretrained models [semantickitti, nuscenes (future)] or load from the provided path (default=`semantickitti`)
+- `--output_rgb_visualization_topic OUTPUT_RGB_VISUALIZATION_TOPIC`: publish the rgb visualization on this topic on `OUTPUT_RGB_VISUALIZATION_TOPIC` (default=`/opendr/panoptic`)
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/README.md:389]: Since this node has point cloud input it should be moved to the relevant point cloud input section. Nodes are separated by input type.

Moreover, documentation for the new node should follow the format in other nodes as closely as possible. Need to add a description at the start with all relevant links as in other nodes, and format the rest as instructions for basic usage steps, please refer  to other nodes' sections.

File: projects/opendr_ws/README.md [hunk 12]
```diff
@@ -73,12 +73,13 @@ Currently, apart from tools, opendr_ws contains the following ROS nodes (categor
...
 6. [2D Single Object Tracking](src/opendr_perception/README.md#2d-single-object-tracking-ros-node)
 7. [2D Object Tracking](src/opendr_perception/README.md#2d-object-tracking-ros-nodes)
-8. [Panoptic Segmentation](src/opendr_perception/README.md#panoptic-segmentation-ros-node)
-9. [Semantic Segmentation](src/opendr_perception/README.md#semantic-segmentation-ros-node)
-10. [Image-based Facial Emotion Estimation](src/opendr_perception/README.md#image-based-facial-emotion-estimation-ros-node)
-11. [Landmark-based Facial Expression Recognition](src/opendr_perception/README.md#landmark-based-facial-expression-recognition-ros-node)
-12. [Skeleton-based Human Action Recognition](src/opendr_perception/README.md#skeleton-based-human-action-recognition-ros-node)
-13. [Video Human Activity Recognition](src/opendr_perception/README.md#video-human-activity-recognition-ros-node)
+8. [Vision Based Panoptic Segmentation](src/opendr_perception/README.md#vision-based-panoptic-segmentation-ros-node)
+9. [LiDAR Based Panoptic Segmentation](src/opendr_perception/README.md#lidar-based-panoptic-segmentation-ros-node)
```
Comment (tsampazk) [inline projects/opendr_ws/README.md:77]: Please, revert the changes in this list and move the new node entry to point cloud input node list below.

File: projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+                # Publish the visualization
+                self._visualization_pointcloud_publisher.publish(ros_pointcloud2_msg)
+
+        except Exception as e:
+            rospy.logwarn(f'Failed to generate prediction: {e}')
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--input_pcl_topic', type=str, default='/usb_cam/pcl_raw',
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py:173]: To be consistent with other nodes that have point cloud input, please refer to argument naming [here](https://github.com/opendr-eu/opendr/blob/9442d1c3f10062b21d71f2f8800841a48fee9c19/projects/opendr_ws/src/opendr_perception/scripts/object_detection_3d_voxel_node.py#L97). You can use something like `--input_point_cloud_2_topic`.

Moreover, default value should be plug and play for the most expected first-time usage. Since you are instructing the user to use the point cloud publisher you provided, this topic value should default to the topic that node publishes, namely `/opendr/dataset_point_cloud2`.

File: projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py [hunk 14]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py:1]: This is required to run smoothly on embedded devices.
```suggestion
#!/usr/bin/env python3
```

File: projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+
+        return joint.ravel().view(dtype)
+
+    def listen(self):
+        """
+        Start the node and begin processing input data. The order of the function calls ensures that the node does not
+        try to process input point clouds without being in a trained state.
+        """
+
+        rospy.init_node("efficient_lps", anonymous=True)
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py:134]:
```suggestion
        rospy.init_node("opendr_efficient_lps_node", anonymous=True)
```

File: projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py [hunk 16]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+# Copyright 2020-2023 OpenDR European Project
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py:2]: This file should be renamed to `panoptic_segmentation_efficient_lps_node.py`.

File: projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
+# Copyright 2020-2023 OpenDR European Project
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py:1]: File should be renamed to `point_cloud_2_publisher_node.py`.

Also shebang:
```suggestion
#!/usr/bin/env python3
# Copyright 2020-2023 OpenDR European Project
```

File: projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+                output_point_cloud_2_topic, ROS_PointCloud2, queue_size=10
+            )
+
+    def start(self):
+        """
+        Starts the node
+        """
+        if self._init_dataset():
+            rospy.loginfo("Starting point_cloud_2 dataset node")
+            rospy.init_node('opendr_point_cloud_2_dataset')
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/point_cloud_2_publisher.py:58]:
```suggestion
            rospy.init_node('opendr_point_cloud_2_dataset_node')
```

File: projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py [hunk 19]
```diff
@@ -0,0 +1,186 @@
...
+        return joint.ravel().view(dtype)
+
+    def listen(self):
+        """
+        Start the node and begin processing input data. The order of the function calls ensures that the node does not
+        try to process input point clouds without being in a trained state.
+        """
+
+        rospy.init_node("efficient_lps", anonymous=True)
+        rospy.loginfo("EfficientLPS node started!")
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py:135]:
```suggestion
        rospy.loginfo("EfficientLPS node started.")
```

**Review by tsampazk — CHANGES_REQUESTED** [review]
Thank you for the new tool! I have added some comments regarding the ROS nodes and the docs. Is there a plan for ROS2?

File: projects/opendr_ws/README.md [hunk 20]
```diff
@@ -49,8 +49,8 @@ For the initial setup you can follow the instructions below:
    You can now run an OpenDR ROS node. More information below.
 
 #### After first time setup
-For running OpenDR nodes after you have completed the initial setup, you can skip step 0 from the list above.
-You can also skip building the workspace (step 4) granted it's been already built and no changes were made to the code inside the workspace, e.g. you modified the source code of a node.
+For running OpenDR nodes after you have completed the initial setup, you can skip steps 2 and 5 from the list above.
+You can also skip building the workspace (step 6) granted it's been already built and no changes were made to the code inside the workspace, e.g. you modified the source code of a node.
```
Comment (tsampazk) [inline projects/opendr_ws/README.md:53]: This last commit [623c80a](https://github.com/opendr-eu/opendr/pull/359/commits/623c80ae46e67f4dd6bd7fb3ca42b6d560afa02f) seems to revert this section to an older version. Did you apply the fixes from an outdated branch maybe? Similarly, on the list below some node  entries are removed.

Comment (aselimc) [reply projects/opendr_ws/README.md:53]: I think I have reverted the file from develop branch. I will check it now again and fix the text. Interesting git stuff happening :D

File: projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py (hunk 13 above)
Comment (aselimc) [reply projects/opendr_ws/src/opendr_perception/scripts/panoptic_segmentation_efficient_lps.py]: Can you please re-check if this in a format that is acceptable?

File: projects/opendr_ws/src/opendr_perception/README.md (hunk 11 above)
Comment (aselimc) [reply projects/opendr_ws/src/opendr_perception/README.md]: I have moved this at the bottom of point cloud inputs. I think this should be sufficient enough at the moment. Sorry for the earlier mess, I didn't realize they are categorized according to their input types.

**Comment by aselimc** [comment]
> Thank you for the new tool! I have added some comments regarding the ROS nodes and the docs. ~Is there a plan for ROS2?~ Edit: Sorry missed the other PR that adds the ROS2 nodes. Note that the ROS1 node and doc comments from this review, might also apply for ROS2.

@tsampazk I just have checked the other PR, and it seems that branch is super old at the moment. Is it okay for me to reset that branch into the current develop and re-write the ROS2 node again?

**Comment by tsampazk** [comment]
> > Thank you for the new tool! I have added some comments regarding the ROS nodes and the docs. ~Is there a plan for ROS2?~ Edit: Sorry missed the other PR that adds the ROS2 nodes. Note that the ROS1 node and doc comments from this review, might also apply for ROS2.
> 
> @tsampazk I just have checked the other PR, and it seems that branch is super old at the moment. Is it okay for me to reset that branch into the current develop and re-write the ROS2 node again?

Sure, go ahead and do it the way you think it's more convenient, it is still draft anyway. :)

File: projects/opendr_ws/src/opendr_perception/README.md (hunk 11 above)
Comment (tsampazk) [reply projects/opendr_ws/src/opendr_perception/README.md]: Thanks, no problem!

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 21]
```diff
@@ -847,3 +876,21 @@ The following optional arguments are available:
    - `-f or --fps FPS`: data fps (default=`10`)
    - `-d or --dataset_path DATASET_PATH`: path to a dataset, if it does not exist, nano KITTI dataset will be downloaded there (default=`/KITTI/opendr_nano_kitti`)
    - `-ks or --kitti_subsets_path KITTI_SUBSETS_PATH`: path to KITTI subsets, used only if a KITTI dataset is downloaded (default=`../../src/opendr/perception/object_detection_3d/datasets/nano_kitti_subsets`)
+
+### Point Cloud 2 Publisher ROS Node
+
+The point cloud 2 dataset publisher, publishes point cloud 2 messages from pre-downloaded dataset SemanticKITTI. It is currently being used by the ROS node [LiDAR Based Panoptic Segmentation ROS Node](#lidar-based-panoptic-segmentation-ros-node)
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/README.md:882]:
```suggestion
The point cloud 2 dataset publisher, publishes point cloud 2 messages from pre-downloaded dataset SemanticKITTI. It is currently being used by the ROS node [LiDAR Based Panoptic Segmentation ROS Node](#lidar-based-panoptic-segmentation-ros-node).
```

File: projects/opendr_ws/src/opendr_perception/README.md [hunk 22]
```diff
@@ -354,11 +354,12 @@ whose documentation can be found here: [Deep Sort docs](../../../../docs/referen
 An [image dataset node](#image-dataset-ros-node) is also provided to be used along these nodes.
 Make sure to change the default input topic of the tracking node if you are not using the USB cam node.
 
-### Panoptic Segmentation ROS Node
+### Vision Based Panoptic Segmentation ROS Node
+A ROS node for performing panoptic segmentation on a specified RGB image stream using the [EfficientPS](../../../../src/opendr/perception/panoptic_segmentation/README.md) network.
```
Comment (tsampazk) [inline projects/opendr_ws/src/opendr_perception/README.md:358]: A direct link to the section should be clearer for someone navigating there.
```suggestion
A ROS node for performing panoptic segmentation on a specified RGB image stream using the [EfficientPS](../../../../src/opendr/perception/panoptic_segmentation/README.md#efficientps-efficient-panoptic-segmentation) network.
```

**Review by tsampazk — CHANGES_REQUESTED** [review]
Thank you, nice job on the fixes. I added some additional very minor comments and should be good to go.

**Comment by aselimc** [comment]
Sorry I had to send an empty commit due to the fact that previous commit wasn't included in the PR somehow, so I had to re-trigger.

**Comment by tsampazk** [comment]
Thank you for the fixes! 

To be honest i haven't been able to test the node myself because it requires the dataset which is quite large. From previous experience there are lots of things that can go wrong when running the nodes that cannot be anticipated beforehand. Is there a possibility of adding a single point cloud and its label on our FTP server for testing purposes similar to other tools, both for the nodes and the demos?

**Comment by aselimc** [comment]
> Thank you for the fixes!
> 
> To be honest i haven't been able to test the node myself because it requires the dataset which is quite large. From previous experience there are lots of things that can go wrong when running the nodes that cannot be anticipated beforehand. Is there a possibility of adding a single point cloud and its label on our FTP server for testing purposes similar to other tools, both for the nodes and the demos?

I will try to have a look on this. When I was testing, I was also using a single point cloud obviously but I need to have closer look to upload this on the FTP.

**Comment by tsampazk** [comment]
> I will try to have a look on this. When I was testing, I was also using a single point cloud obviously but I need to have closer look to upload this on the FTP.

Yes of course, if there are licensing or other issues we can work with what we have. :smile:

**Comment by tsampazk** [comment]
Thank you very much @aselimc! 

Regarding the download methods, for consistency reasons and for being easier to use (e.g. through demos too), i would suggest moving the download stuff inside the learner's download method, by adding a second `mode`  similar to [other nodes](https://github.com/opendr-eu/opendr/blob/master/src/opendr/perception/pose_estimation/lightweight_open_pose/lightweight_open_pose_learner.py#L917), sorry for not making more clear earlier.

I tested the node and unless i'm mistaken, it seems that the output topic is not really an RGB visualization topic as the name suggests. Similar to the other panoptic segmentation node, there are two output topics expected, one where the heatmap is published and another one where the heatmap is blended with an RGB image for visualization purposes.

**Comment by aselimc** [comment]
> Thank you very much @aselimc!
> 
> Regarding the download methods, for consistency reasons and for b eing easier to use (e.g. through demos too), i would suggest moving the download stuff inside the learner's download method, by adding a second `mode` similar to [other nodes](https://github.com/opendr-eu/opendr/blob/master/src/opendr/perception/pose_estimation/lightweight_open_pose/lightweight_open_pose_learner.py#L917), sorry for not making more clear earlier.
> 
> I tested the node and unless i'm mistaken, it seems that the output topic is not really an RGB visualization topic as the name suggests. Similar to the other panoptic segmentation node, there are two output topics expected, one where the heatmap is published and another one where the heatmap is blended with an RGB image for visualization purposes.

@tsampazk For the first suggestion, we already have the download method of download but it was a bit confusing for me to call the learner in two nodes due to the error that I mentioned earlier. But now I figured out that my environment was the root cause for the error so I will fix this.

But about the second point, I believe the purpose of panoptic segmentation is to simply merge the semantic and instance segmentations and publish according to this. Thus, the RGB visualization topic publishes an RGB labelled pointcloud where users can visualize it with tools like RVIZ. I can separate the panoptic output of pointcloud into semantic and instance segmentation maps but I think this does not bring much to the table. @vniclas What do you suggest? Also, I am not entirely sure about heatmap+input combination for this task simply because inputs are not RGB images.

**Comment by tsampazk** [comment]
> @tsampazk For the first suggestion, we already have the download method of download but it was a bit confusing for me to call the learner in two nodes due to the error that I mentioned earlier. But now I figured out that my environment was the root cause for the error so I will fix this.

It is true that it doesn't make much sense to initialize and call the learner in the point cloud publisher, so i would suggest to keep that as-is and add a `test_data` mode to the learner's download method that basically just downloads the data. This will make it possible to quickly run the demos of the tool. This will lead to some minor code duplication but i think it's acceptable.

>But about the second point, I believe the purpose of panoptic segmentation is to simply merge the semantic and instance segmentations and publish according to this. Thus, the RGB visualization topic publishes an RGB labelled pointcloud where users can visualize it with tools like RVIZ. I can separate the panoptic output of pointcloud into semantic and instance segmentation maps but I think this does not bring much to the table. @vniclas What do you suggest? Also, I am not entirely sure about heatmap+input combination for this task simply because inputs are not RGB images.

You're right, when looking at the output i missed it entirely that this node indeed doesn't work with RGB stuff :confused:. 

The output topic doesn't show up on rqt-image-viewer which i have been using for node testing to do visual verification. I test the nodes this way because from use-case partner feedback in other nodes, they usually first expect something tangible in the output, i.e. visual output, and secondly a detection, e.g. a heatmap or a bounding box which can be echoed. If it is indeed not meaningful or possible to have an easy visualization of the output then i suggest renaming the topic to something that doesn't indeed imply (easy) RGB visualization to avoid confusion.

I'm sorry about the wall of text and insisting on the matter, we had several user issues in the past about the output topics.

**Comment by aselimc** [comment]
> > @tsampazk For the first suggestion, we already have the download method of download but it was a bit confusing for me to call the learner in two nodes due to the error that I mentioned earlier. But now I figured out that my environment was the root cause for the error so I will fix this.
> 
> It is true that it doesn't make much sense to initialize and call the learner in the point cloud publisher, so i would suggest to keep that as-is and add a `test_data` mode to the learner's download method that basically just downloads the data. This will make it possible to quickly run the demos of the tool. This will lead to some minor code duplication but i think it's acceptable.
> 
> > But about the second point, I believe the purpose of panoptic segmentation is to simply merge the semantic and instance segmentations and publish according to this. Thus, the RGB visualization topic publishes an RGB labelled pointcloud where users can visualize it with tools like RVIZ. I can separate the panoptic output of pointcloud into semantic and instance segmentation maps but I think this does not bring much to the table. @vniclas What do you suggest? Also, I am not entirely sure about heatmap+input combination for this task simply because inputs are not RGB images.
> 
> You're right, when looking at the output i missed it entirely that this node indeed doesn't work with RGB stuff confused.
> 
> The output topic doesn't show up on rqt-image-viewer which i have been using for node testing to do visual verification. I test the nodes this way because from use-case partner feedback in other nodes, they usually first expect something tangible in the output, i.e. visual output, and secondly a detection, e.g. a heatmap or a bounding box which can be echoed. If it is indeed not meaningful or possible to have an easy visualization of the output then i suggest renaming the topic to something that doesn't indeed imply (easy) RGB visualization to avoid confusion.
> 
> I'm sorry about the wall of text and insisting on the matter, we had several user issues in the past about the output topics.

Sorry for the late reply. Right now the download method is a static method where it is called from the learner and I have further adjusted the naming of the output topic as you have suggested. Please let me know if there are any extra changes required.

**Review by tsampazk — APPROVED** [review]
Thank you @aselimc for your efforts! :rocket:

**Comment by vniclas** [comment]
Thanks for the extensive review @tsampazk :pray: 

@passalis , could you please also have a look when you have time so that we can merge this PR

**Review by ad-daniel — APPROVED** [review]
All my points have been addressed, thank you!
